src/test-explorer/test-manager.ts

```diff
@@ -0,0 +1,163 @@
+import * as vscode from "vscode";
+import { TestRunProfileKind, tests } from "vscode";
+import {
+  ExecuteCommandRequest,
+  LanguageClient,
+} from "vscode-languageclient/node";
+import { testCache } from "./test-cache";
+import { runHandler } from "./test-run-handler";
+import {
+  SuiteDiscovery,
+  TargetName,
+  TargetUri,
+  TestDiscoveryResult,
+  TestItemMetadata,
+} from "./types";
+import { toVscodeRange } from "./util";
+
+export function createTestManager(
+  client: LanguageClient,
+  isDisabled: boolean
+): TestManager {
+  return new TestManager(client, isDisabled);
+}
+
+class TestManager {
+  readonly testController = tests.createTestController(
+    "metalsTestController",
+    "Metals Test Explorer"
+  );
+
+  private isDisabled = false;
+
+  constructor(private readonly client: LanguageClient, isDisabled: boolean) {
+    if (isDisabled) {
+      this.disable();
+    }
+    this.testController.resolveHandler = async (item?: vscode.TestItem) => {
+      if (item != null) {
+        return;
+      } else {
+        await this.discoverTestSuites();
+      }
+    };
+
+    this.testController.createRunProfile(
+      "Run",
+      TestRunProfileKind.Run,
+      (request, token) => runHandler(this.testController, true, request, token),
+      true
+    );
+
+    this.testController.createRunProfile(
+      "Debug",
+      TestRunProfileKind.Debug,
+      (request, token) =>
+        runHandler(this.testController, false, request, token),
+      false
+    );
+  }
+
+  async enable(): Promise<void> {
+    this.isDisabled = false;
+    await this.discoverTestSuites();
+  }
+
+  /**
+   * Disables test manager, also it deletes all discovered test items in order to hide gutter icons.
+   */
+  disable(): void {
+    this.testController.items.forEach((item) =>
+      this.testController.items.delete(item.id)
```

Comment:
Should we also empty the TestCache?

Reply:
Actually, there is no point in doing that. `this.testController.items.forEach` iterates only over top-level test items which are projects. In order to delete everything we will have to extract all the children from projects and it's not worth it.
Old test items will be eventually GC'ed or even replaced by the newest ones.